File: ui/prometheus-plugin/src/plugins/prometheus-time-series-query/plugin.test.ts
// Copyright 2023 The Perses Authors
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// TODO: This should be fixed globally in the test setup
import { DatasourceSpec } from '@perses-dev/core';

jest.mock('echarts/core');

import { TimeSeriesQueryContext } from '@perses-dev/plugin-system';
import { RangeQueryResponse } from '../../model';
import { PrometheusDatasource } from '../prometheus-datasource';
import { PrometheusDatasourceSpec } from '../types';
import { PrometheusTimeSeriesQuery } from './';

const datasource: PrometheusDatasourceSpec = {
  directUrl: '/test',
  scrapeInterval: '1m',
};

const promStubClient = PrometheusDatasource.createClient(datasource, {});

// Mock range query
promStubClient.rangeQuery = jest.fn(async () => {
  const stubRepsonse: RangeQueryResponse = {
    status: 'success',
    data: {
      resultType: 'matrix',
      result: [
        {
          metric: {
            __name__: 'up',
          },
          values: [[1686141338.877, '10']],
        },
      ],
    },
  };
  return stubRepsonse;
});

const getDatasourceClient: jest.Mock = jest.fn(() => {
  return promStubClient;
});

const getDatasource: jest.Mock = jest.fn((): DatasourceSpec<PrometheusDatasourceSpec> => {
  return {
    default: false,
    plugin: {
      kind: 'PrometheusDatasource',
      spec: datasource,
    },
  };
});

const createStubContext = (): TimeSeriesQueryContext => {
  const stubTimeSeriesContext: TimeSeriesQueryContext = {
    datasourceStore: {
      getDatasource: getDatasource,
      getDatasourceClient: getDatasourceClient,
      listDatasourceSelectItems: jest.fn(),
      getLocalDatasources: jest.fn(),
      setLocalDatasources: jest.fn(),
      getSavedDatasources: jest.fn(),
      setSavedDatasources: jest.fn(),
    },
    refreshKey: 'test',
    timeRange: {
      end: new Date('01-01-2023'),
      start: new Date('01-02-2023'),
    },
    variableState: {},
  };
  return stubTimeSeriesContext;
};

describe('PrometheusTimeSeriesQuery', () => {
  it('should properly resolve variable dependencies', () => {
    if (!PrometheusTimeSeriesQuery.dependsOn) throw new Error('dependsOn is not defined');
    const { variables } = PrometheusTimeSeriesQuery.dependsOn(
      {
        query: 'sum(up{job="$job"}) by ($instance)',
        seriesNameFormat: `$foo - label`,
      },
      createStubContext()
    );
    expect(variables).toEqual(['job', 'instance', 'foo']);
  });

  it('should replace variables in seriesNameFormat', async () => {
    const ctx = createStubContext();
    ctx.variableState = {
      foo: {
        value: 'bar',
        loading: false,
      },
    };

    const results = await PrometheusTimeSeriesQuery.getTimeSeriesData(
      {
        query: 'sum(up{job="$job"}) by ($instance)',
        seriesNameFormat: `$foo - format`,
      },
      ctx
    );

    expect(results.series[0]?.formattedName).toEqual('bar - format');
  });
});

File: internal/api/plugin/plugin_test.go
// Copyright The Perses Authors
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package plugin

import (
	"testing"

	v1 "github.com/perses/perses/pkg/model/api/v1"
	"github.com/perses/spec/go/module"
	"github.com/perses/spec/go/plugin"
	"github.com/stretchr/testify/assert"
)

func TestFilter(t *testing.T) {
	tests := []struct {
		title          string
		enabled        []string
		disabled       []string
		module         *v1.PluginModule
		isFiltered     bool
		expectedModule *v1.PluginModule
	}{
		{
			title: "no filter",
			module: &v1.PluginModule{
				Kind: v1.PluginModuleKind,
				Metadata: module.Metadata{
					Name: "foo",
				},
				Spec: v1.ModuleSpec{
					Plugins: []module.Plugin{
						{
							Kind: plugin.KindPanel,
							Spec: module.PluginSpec{
								Name: "foo",
							},
						},
					},
				},
			},
			isFiltered: false,
			expectedModule: &v1.PluginModule{
				Kind: v1.PluginModuleKind,
				Metadata: module.Metadata{
					Name: "foo",
				},
				Spec: v1.ModuleSpec{
					Plugins: []module.Plugin{
						{
							Kind: plugin.KindPanel,
							Spec: module.PluginSpec{
								Name: "foo",
							},
						},
					},
				},
			},
		},
		{
			title: "activated",
			enabled: []string{
				"foo",
			},
			module: &v1.PluginModule{
				Kind: v1.PluginModuleKind,
				Metadata: module.Metadata{
					Name: "foo",
				},
				Spec: v1.ModuleSpec{
					Plugins: []module.Plugin{
						{
							Kind: plugin.KindPanel,
							Spec: module.PluginSpec{
								Name: "foo",
							},
						},
					},
				},
			},
			isFiltered: false,
			expectedModule: &v1.PluginModule{
				Kind: v1.PluginModuleKind,
				Metadata: module.Metadata{
					Name: "foo",
				},
				Spec: v1.ModuleSpec{
					Plugins: []module.Plugin{
						{
							Kind: plugin.KindPanel,
							Spec: module.PluginSpec{
								Name: "foo",
							},
						},
					},
				},
			},
		},
		{
			title: "not activated",
			enabled: []string{
				"bar",
			},
			module: &v1.PluginModule{
				Kind: v1.PluginModuleKind,
				Metadata: module.Metadata{
					Name: "foo",
				},
				Spec: v1.ModuleSpec{
					Plugins: []module.Plugin{
						{
							Kind: plugin.KindPanel,
							Spec: module.PluginSpec{
								Name: "foo",
							},
						},
					},
				},
			},
			isFiltered: true,
			expectedModule: &v1.PluginModule{
				Kind: v1.PluginModuleKind,
				Metadata: module.Metadata{
					Name: "foo",
				},
				Spec: v1.ModuleSpec{},
			},
		},
		{
			title: "activated but with less plugins",
			enabled: []string{
				"bar",
			},
			module: &v1.PluginModule{
				Kind: v1.PluginModuleKind,
				Metadata: module.Metadata{
					Name: "foo",
				},
				Spec: v1.ModuleSpec{
					Plugins: []module.Plugin{
						{
							Kind: plugin.KindPanel,
							Spec: module.PluginSpec{
								Name: "tartampion",
							},
						},
						{
							Kind: plugin.KindPanel,
							Spec: module.PluginSpec{
								Name: "bar",
							},
						},
					},
				},
			},
			isFiltered: false,
			expectedModule: &v1.PluginModule{
				Kind: v1.PluginModuleKind,
				Metadata: module.Metadata{
					Name: "foo",
				},
				Spec: v1.ModuleSpec{
					Plugins: []module.Plugin{
						{
							Kind: plugin.KindPanel,
							Spec: module.PluginSpec{
								Name: "bar",
							},
						},
					},
				},
			},
		},
		{
			title: "deactivated",
			disabled: []string{
				"foo",
			},
			module: &v1.PluginModule{
				Kind: v1.PluginModuleKind,
				Metadata: module.Metadata{
					Name: "foo",
				},
				Spec: v1.ModuleSpec{
					Plugins: []module.Plugin{
						{
							Kind: plugin.KindPanel,
							Spec: module.PluginSpec{
								Name: "bar",
							},
						},
					},
				},
			},
			isFiltered: true,
			expectedModule: &v1.PluginModule{
				Kind: v1.PluginModuleKind,
				Metadata: module.Metadata{
					Name: "foo",
				},
				Spec: v1.ModuleSpec{
					Plugins: []module.Plugin{
						{
							Kind: plugin.KindPanel,
							Spec: module.PluginSpec{
								Name: "bar",
							},
						},
					},
				},
			},
		},
		{
			title: "deactivated with no plugins",
			disabled: []string{
				"bar",
			},
			module: &v1.PluginModule{
				Kind: v1.PluginModuleKind,
				Metadata: module.Metadata{
					Name: "foo",
				},
				Spec: v1.ModuleSpec{
					Plugins: []module.Plugin{
						{
							Kind: plugin.KindPanel,
							Spec: module.PluginSpec{
								Name: "bar",
							},
						},
					},
				},
			},
			isFiltered: true,
			expectedModule: &v1.PluginModule{
				Kind: v1.PluginModuleKind,
				Metadata: module.Metadata{
					Name: "foo",
				},
				Spec: v1.ModuleSpec{},
			},
		},
		{
			title: "deactivated with less plugins",
			disabled: []string{
				"bar",
			},
			module: &v1.PluginModule{
				Kind: v1.PluginModuleKind,
				Metadata: module.Metadata{
					Name: "foo",
				},
				Spec: v1.ModuleSpec{
					Plugins: []module.Plugin{
						{
							Kind: plugin.KindPanel,
							Spec: module.PluginSpec{
								Name: "bar",
							},
						},
						{
							Kind: plugin.KindPanel,
							Spec: module.PluginSpec{
								Name: "tartampion",
							},
						},
					},
				},
			},
			isFiltered: false,
			expectedModule: &v1.PluginModule{
				Kind: v1.PluginModuleKind,
				Metadata: module.Metadata{
					Name: "foo",
				},
				Spec: v1.ModuleSpec{
					Plugins: []module.Plugin{
						{
							Kind: plugin.KindPanel,
							Spec: module.PluginSpec{
								Name: "tartampion",
							},
						},
					},
				},
			},
		},
		{
			title: "not deactivated",
			disabled: []string{
				"tartampion",
			},
			module: &v1.PluginModule{
				Kind: v1.PluginModuleKind,
				Metadata: module.Metadata{
					Name: "foo",
				},
				Spec: v1.ModuleSpec{
					Plugins: []module.Plugin{
						{
							Kind: plugin.KindPanel,
							Spec: module.PluginSpec{
								Name: "bar",
							},
						},
					},
				},
			},
			isFiltered: false,
			expectedModule: &v1.PluginModule{
				Kind: v1.PluginModuleKind,
				Metadata: module.Metadata{
					Name: "foo",
				},
				Spec: v1.ModuleSpec{
					Plugins: []module.Plugin{
						{
							Kind: plugin.KindPanel,
							Spec: module.PluginSpec{
								Name: "bar",
							},
						},
					},
				},
			},
		},
	}
	for _, test := range tests {
		t.Run(test.title, func(t *testing.T) {
			p := &pluginFile{
				enabled:  test.enabled,
				disabled: test.disabled,
			}
			result := p.filter(test.module)
			assert.Equal(t, test.expectedModule, test.module)
			assert.Equal(t, test.isFiltered, result)
		})
	}
}
